File: ios-native/ClubFuoco/Models/Social.swift
import Foundation

// ── Friends (GET /api/friends, /api/friends/search) ──────────────────────────

struct FriendUser: Decodable, Identifiable, Sendable, Hashable {
    let id: UUID
    let fullName: String?
    let avatarUrl: String?
    let friendshipId: UUID

    var initials: String {
        let parts = (fullName ?? "").split(separator: " ")
        let chars = parts.prefix(2).compactMap { $0.first.map(String.init) }
        return chars.isEmpty ? "?" : chars.joined().uppercased()
    }
}

struct FriendsData: Decodable, Sendable {
    let friends: [FriendUser]
    let incoming: [FriendUser]
    let outgoing: [FriendUser]
}

struct FriendSearchResult: Decodable, Identifiable, Sendable {
    let id: UUID
    let fullName: String?
    let avatarUrl: String?
    let relation: String        // none | friends | outgoing | incoming
    let friendshipId: UUID?

    var initials: String {
        let parts = (fullName ?? "").split(separator: " ")
        let chars = parts.prefix(2).compactMap { $0.first.map(String.init) }
        return chars.isEmpty ? "?" : chars.joined().uppercased()
    }
}

// ── Groups (GET /api/groups, /api/groups/[id]) ────────────────────────────────

struct GroupListItem: Decodable, Identifiable, Sendable, Hashable {
    struct ClubInfo: Decodable, Sendable, Hashable {
        let name: String?
        let coverImageUrl: String?
    }

    let id: UUID
    let bookingType: String
    let bookingDate: String
    let status: String
    let inviteCode: String?
    let clubs: EmbeddedOne<ClubInfo>?
    let myRsvp: String?

    var club: ClubInfo? { clubs?.value }
}

struct GroupMember: Decodable, Identifiable, Sendable {
    let id: UUID
    let userId: UUID
    let fullName: String?
    let avatarUrl: String?
    let role: String            // organizer | member
    let rsvp: String            // invited | going | maybe | declined
    let paymentRequired: Bool
    let amountDue: Double?
    let charge: Double
    let paid: Bool
    let isMe: Bool
    let bookingId: UUID?        // this member's own entry booking (only set for me)
    let qrToken: String?        // their pass QR (only set for me — never others')

    var initials: String {
        let parts = (fullName ?? "").split(separator: " ")
        let chars = parts.prefix(2).compactMap { $0.first.map(String.init) }
        return chars.isEmpty ? "?" : chars.joined().uppercased()
    }
}

struct GroupDetail: Decodable, Sendable {
    let id: UUID
    let clubId: UUID
    let clubName: String
    let clubImage: String?
    let organizerId: UUID
    let organizerName: String?
    let bookingType: String
    let bookingDate: String
    let inviteCode: String
    let status: String
    let unitPrice: Double
    let members: [GroupMember]
    let me: GroupMember?
}

struct GroupJoinResult: Decodable, Sendable {
    let rsvp: String
    let bookingId: UUID?
    let qrToken: String?
}

// ── Rumbas (GET /api/rumbas, signups) ─────────────────────────────────────────

struct Rumba: Decodable, Identifiable, Sendable, Hashable {
    let id: UUID
    let title: String
    let venueName: String?
    let venuePlaceId: String?
    let eventDate: String?      // timestamp string
    let capacity: Int?
    let description: String?
    let coverImage: String?
    let dressCode: String?
    let signupCount: Int?

    var spotsLeft: Int {
        max(0, (capacity ?? 0) - (signupCount ?? 0))
    }
}

struct RumbaSignup: Decodable, Sendable {
    let id: UUID
    let status: String?         // confirmed | waitlist | denied …
    let plusOnes: Int?
    let name: String?
}

// ── Fiamme (GET /api/fiamme) ──────────────────────────────────────────────────

struct FiammeData: Decodable, Sendable {
    struct Activity: Decodable, Identifiable, Sendable {
        let id: UUID
        let amount: Int
        let type: String?
        let description: String?
        let createdAt: String?
    }

    let balance: Int
    let reviewCount: Int
    let activity: [Activity]
}

struct FiammeRedeemResult: Decodable, Sendable {
    let code: String
}

/// Fixed reward catalogue (mirrors /api/fiamme/redeem REWARDS).
struct FiammeReward: Identifiable {
    let key: String
    let cost: Int
    var id: String { key }

    static let all: [FiammeReward] = [
        FiammeReward(key: "comp_drink", cost: 100),
        FiammeReward(key: "skip_line", cost: 250),
        FiammeReward(key: "free_cover", cost: 500),
        FiammeReward(key: "bottle_deposit", cost: 1000),
    ]
}

/// Fiamme tier ladder (mirrors TIERS in fiamme/page.tsx; thresholds = reviews).
struct FiammeTier {
    let key: String
    let label: String           // brand (Italian) name
    let min: Int
    let mult: String

    static let all: [FiammeTier] = [
        FiammeTier(key: "nuovo", label: "Nuovo", min: 0, mult: "1×"),
        FiammeTier(key: "regolare", label: "Regolare", min: 5, mult: "1.2×"),
        FiammeTier(key: "conoscitore", label: "Conoscitore", min: 15, mult: "1.5×"),
        FiammeTier(key: "tastemaker", label: "Tastemaker", min: 30, mult: "2×"),
    ]

    static func current(reviews: Int) -> FiammeTier {
        all.last { reviews >= $0.min } ?? all[0]
    }
}

// ── Notifications (GET /api/notifications) ────────────────────────────────────

struct NotificationItem: Decodable, Identifiable, Sendable {
    let id: UUID
    let type: String?
    let title: String
    let body: String?
    let link: String?
    let isRead: Bool?
    let createdAt: String?
}

File: android-native/app/src/main/java/com/clubfuoco/app/models/Social.kt
package com.clubfuoco.app.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ── Friends (GET /api/friends, /api/friends/search) ──────────────────────────

/** Initials for the avatar placeholder — first letters of the first two words. */
private fun initialsOf(fullName: String?): String {
    val chars = fullName.orEmpty().split(" ")
        .filter { it.isNotEmpty() }
        .take(2)
        .mapNotNull { it.firstOrNull() }
    return if (chars.isEmpty()) "?" else chars.joinToString("").uppercase()
}

@Serializable
data class FriendUser(
    val id: String,
    @SerialName("fullName") val fullName: String? = null,
    @SerialName("avatarUrl") val avatarUrl: String? = null,
    @SerialName("friendshipId") val friendshipId: String? = null,
) {
    val initials: String get() = initialsOf(fullName)
}

@Serializable
data class FriendsData(
    val friends: List<FriendUser> = emptyList(),
    val incoming: List<FriendUser> = emptyList(),
    val outgoing: List<FriendUser> = emptyList(),
)

@Serializable
data class FriendSearchResult(
    val id: String,
    @SerialName("fullName") val fullName: String? = null,
    @SerialName("avatarUrl") val avatarUrl: String? = null,
    /** none | friends | outgoing | incoming */
    val relation: String = "none",
    @SerialName("friendshipId") val friendshipId: String? = null,
) {
    val initials: String get() = initialsOf(fullName)
}

// ── Notifications (GET /api/notifications) ───────────────────────────────────

@Serializable
data class AppNotification(
    val id: String,
    val type: String? = null,
    val title: String? = null,
    val body: String? = null,
    @SerialName("is_read") val isRead: Boolean = false,
    @SerialName("created_at") val createdAt: String? = null,
)

// ── Fiamme (GET /api/fiamme) ─────────────────────────────────────────────────

/**
 * The loyalty balance. Field names follow the REST payload rather than the
 * database, because this one is served by the API route, not PostgREST.
 */
@Serializable
data class FiammeData(
    val balance: Int = 0,
    val tier: String? = null,
    @SerialName("nextTier") val nextTier: String? = null,
    @SerialName("toNextTier") val toNextTier: Int? = null,
    val activity: List<FiammeEntry> = emptyList(),
)

@Serializable
data class FiammeEntry(
    val id: String,
    val amount: Int = 0,
    val reason: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
)

// ── Groups (GET /api/groups, /api/groups/[id]) ───────────────────────────────

@Serializable
data class GroupClubInfo(
    val name: String? = null,
    @SerialName("coverImageUrl") val coverImageUrl: String? = null,
)

@Serializable
data class GroupListItem(
    val id: String,
    @SerialName("clubId") val clubId: String? = null,
    @SerialName("bookingType") val bookingType: String = "general",
    @SerialName("bookingDate") val bookingDate: String = "",
    val status: String = "open",
    @SerialName("inviteCode") val inviteCode: String? = null,
    val clubs: GroupClubInfo? = null,
    @SerialName("myRsvp") val myRsvp: String? = null,
) {
    val club: GroupClubInfo? get() = clubs
}

@Serializable
data class GroupMember(
    val id: String,
    @SerialName("userId") val userId: String? = null,
    @SerialName("fullName") val fullName: String? = null,
    @SerialName("avatarUrl") val avatarUrl: String? = null,
    /** organizer | member */
    val role: String = "member",
    /** invited | going | maybe | declined */
    val rsvp: String = "invited",
    @SerialName("paymentRequired") val paymentRequired: Boolean = false,
    @SerialName("amountDue") val amountDue: Double? = null,
    val charge: Double = 0.0,
    val paid: Boolean = false,
    @SerialName("isMe") val isMe: Boolean = false,
    @SerialName("bookingId") val bookingId: String? = null,
    /** Their pass QR — only ever set for me, never for other members. */
    @SerialName("qrToken") val qrToken: String? = null,
) {
    val initials: String get() = initialsOf(fullName)
}

@Serializable
data class GroupDetail(
    val id: String,
    @SerialName("clubId") val clubId: String? = null,
    @SerialName("clubName") val clubName: String = "",
    @SerialName("clubImage") val clubImage: String? = null,
    @SerialName("organizerId") val organizerId: String? = null,
    @SerialName("organizerName") val organizerName: String? = null,
    @SerialName("bookingType") val bookingType: String = "general",
    @SerialName("bookingDate") val bookingDate: String = "",
    @SerialName("inviteCode") val inviteCode: String = "",
    val status: String = "open",
    @SerialName("unitPrice") val unitPrice: Double = 0.0,
    val members: List<GroupMember> = emptyList(),
    val me: GroupMember? = null,
    /** Optional so the detail still decodes against an older deploy; nil = 0. */
    @SerialName("unreadCount") val unreadCount: Int? = null,
)

@Serializable
data class GroupMessage(
    val id: String,
    @SerialName("userId") val userId: String? = null,
    @SerialName("senderName") val senderName: String? = null,
    @SerialName("senderAvatarUrl") val senderAvatarUrl: String? = null,
    val body: String = "",
    @SerialName("createdAt") val createdAt: String = "",
    @SerialName("isMine") val isMine: Boolean = false,
) {
    val initials: String get() = initialsOf(senderName)

    /** First name for the small sender label above other people's bubbles. */
    val shortName: String
        get() = senderName.orEmpty().split(" ").firstOrNull().orEmpty().ifEmpty { "Someone" }
}
